lib/matobjplist.gi

```diff
@@ -71,33 +71,81 @@ BindGlobal( "MakeIsPlistVectorRep",
-  function( filter, basedomain, l )
-    if ValueOption( "check" ) <> false and not IsSubset( basedomain, l ) then
-      Error( "the elements in <l> must lie in <basedomain>" );
+BindGlobal( "MakeIsPlistMatrixRep",
+  function( basedomain, emptyvector, rowlength, list, check )
+    local fam, types, typ, row;
+    fam:= CollectionsFamily( FamilyObj( basedomain ) );
+
+    # Currently there is no special handling depending on 'basedomain',
+    # the types are always cached in 'fam'.
+    if not IsBound( fam!.PlistMatrixRepTypes ) then
+      # initialize type cache
+      # TODO: make this thread safe for HPC-GAP
+      fam!.PlistMatrixRepTypes:= [
+          NewType( fam, IsPlistMatrixRep ),
+          NewType( fam, IsPlistMatrixRep and IsMutable ),
+      ];
+      fam!.PlistMatrixRepTypesEasyCompare:= [
+          NewType( fam, IsPlistMatrixRep and CanEasilyCompareElements ),
+          NewType( fam, IsPlistMatrixRep and CanEasilyCompareElements and IsMutable ),
+      ];
     fi;
-    return MakeIsPlistVectorRep(basedomain, ShallowCopy(l));
+    if HasCanEasilyCompareElements( Representative( basedomain ) ) and
+       CanEasilyCompareElements( Representative( basedomain ) ) then
+      types:= fam!.PlistMatrixRepTypesEasyCompare;
+    else
+      types:= fam!.PlistMatrixRepTypes;
+    fi;
+    if IsMutable( list ) then
+      typ:= types[2];
+    else
+      typ:= types[1];
+    fi;
+
+    if check and ValueOption( "check" ) <> false then
+      for row in list do
+        if not IsPlistVectorRep( row ) then
+          Error( "the entries of <list> must be in 'IsPlistVectorRep'" );
+        elif not IsIdenticalObj( basedomain, row![BDPOS] ) then
+          Error( "the entries of <list> must have the given base domain" );
+        elif not IsSubset( basedomain, row![ELSPOS] ) then
+          Error( "the elements in <row> must lie in <basedomain>" );
+        elif Length( row![ELSPOS] ) <> rowlength then
+          Error( "the entries of <list> must have length <rowlength>" );
+        fi;
+      od;
+    fi;
+
+    return Objectify( typ, [ basedomain, emptyvector, rowlength, list ] );
+  end );
+
+InstallMethod( NewVector,
+  [ "IsPlistVectorRep", "IsRing", "IsDenseList" ],
+  function( filter, basedomain, l )
+    return MakeIsPlistVectorRep(basedomain, ShallowCopy(l), true);
   end );
 
-InstallMethod( NewZeroVector, "for IsPlistVectorRep, a ring, and an int",
-  [ IsPlistVectorRep, IsRing, IsInt ],
+InstallMethod( NewZeroVector,
+  [ "IsPlistVectorRep", "IsRing", "IsInt" ],
   function( filter, basedomain, len )
     local list;
     list := ListWithIdenticalEntries(len, Zero(basedomain));
-    return MakeIsPlistVectorRep(basedomain, list);
+    return MakeIsPlistVectorRep(basedomain, list, false);
   end );
 
 InstallMethod( NewMatrix,
-  "for IsPlistMatrixRep, a ring, an int, and a list",
-  [ IsPlistMatrixRep, IsRing, IsInt, IsList ],
+  [ "IsPlistMatrixRep", "IsRing", "IsInt", "IsList" ],
   function( filter, basedomain, rl, l )
```

Comment:
this is out of scope for this PR, but I would suggest we consistently replace `rl` by `ncols` or `numcols` or something like that (because `row length` is mostly gone now and so it is hard to guess what `rl` stands for). Also, perhaps `l` should be turned into `list`